src/openrct2/core/OrcaStream.hpp

```diff
@@ -0,0 +1,685 @@
+            }
+        }
+
+        OrcaStream(const OrcaStream&) = delete;
+
+        ~OrcaStream()
+        {
+            if (_mode == Mode::READING)
+            {
+            }
+            else
+            {
+                const void* uncompressedData = _buffer.GetData();
+                const uint64_t uncompressedSize = _buffer.GetLength();
+
+                _header.NumChunks = static_cast<uint32_t>(_chunks.size());
+                _header.UncompressedSize = uncompressedSize;
+                _header.CompressedSize = uncompressedSize;
+                _header.FNV1a = Crypt::FNV1a(uncompressedData, uncompressedSize);
+
+                // Compress data
+                std::optional<std::vector<uint8_t>> compressedBytes;
+                if (_header.Compression == COMPRESSION_GZIP)
+                {
+                    compressedBytes = Gzip(uncompressedData, uncompressedSize);
+                    if (compressedBytes)
+                    {
+                        _header.CompressedSize = compressedBytes->size();
+                    }
+                    else
+                    {
+                        // Compression failed
+                        _header.Compression = COMPRESSION_NONE;
+                    }
+                }
+
+                // Write header and chunk table
+                _stream->WriteValue(_header);
+                for (const auto& chunk : _chunks)
+                {
+                    _stream->WriteValue(chunk);
+                }
+
+                // Write chunk data
+                if (compressedBytes)
+                {
+                    _stream->Write(compressedBytes->data(), compressedBytes->size());
+                }
+                else
+                {
+                    _stream->Write(uncompressedData, uncompressedSize);
+                }
+            }
+        }
+
+        Mode GetMode() const
+        {
+            return _mode;
+        }
+
+        Header& GetHeader()
+        {
+            return _header;
+        }
+
+        const Header& GetHeader() const
+        {
+            return _header;
+        }
+
+        template<typename TFunc> bool ReadWriteChunk(uint32_t chunkId, TFunc f)
+        {
+            if (_mode == Mode::READING)
+            {
+                if (SeekChunk(chunkId))
+                {
+                    ChunkStream stream(_buffer, _mode);
+                    f(stream);
+                    return true;
+                }
+
+                return false;
+            }
+
+            _currentChunk.Id = chunkId;
+            _currentChunk.Offset = _buffer.GetPosition();
+            _currentChunk.Length = 0;
+            ChunkStream stream(_buffer, _mode);
+            f(stream);
+            _currentChunk.Length = static_cast<uint64_t>(_buffer.GetPosition()) - _currentChunk.Offset;
+            _chunks.push_back(_currentChunk);
+            return true;
+        }
+
+    private:
+        bool SeekChunk(uint32_t id)
```

Comment:
Can probably be made const.

```suggestion
        bool SeekChunk(uint32_t id) const
```

Reply:
It does mutate the container, I don't think const applies here.

Reply:
Right, it modifies `_buffer`. Wasn't quite sure when looking this over the first time, hence 'probably'.